Change list striding parameters to take optional integer (preceding work for #49352)

tugsbayasgalan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #49352 Striding for lists Part 2
* **#48719 Striding for lists Part 1**

Summary: Attempt to break this PR (https://github.com/pytorch/pytorch/pull/33019) into two parts. As per our discussion with @eellison,  the first part is to make sure our aten::slice operator take optional parameters for begin/step/end. This will help with refactoring ir_emitter.cpp for genering handling for list and slice striding. Once this PR merged, we will submit a second PR with compiler change. 

Test Plan: None for this PR, but new tests will be added for the second part.

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D25929902](https://our.internmc.facebook.com/intern/diff/D25929902)

tugsbayasgalan committed:
Striding for lists

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "[WIP] Striding for lists"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "[WIP] Striding for lists"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "[WIP] Striding for lists"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

albanD reviewed (commented):
What is actually changing here? It would be nice to have the codemod in a separate PR.

eellison commented:
Yea, agreed with alban, it might be easier to review this once the whole stack is out

tugsbayasgalan committed:
Update on "Striding for lists Part 1"


Summary: Attempt to break this PR (https://github.com/pytorch/pytorch/pull/33019) into two parts. As per our discussion with @eellison,  the first part is to make sure our aten::slice operator take optional parameters for begin/step/end. This will help with refactoring ir_emitter.cpp for genering handling for list and slice striding. Once this PR merged, we will submit a second PR with compiler change. 

Test Plan: None for this PR, but new tests will be added for the second part.

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Striding for lists Part 1"


Summary: Attempt to break this PR (https://github.com/pytorch/pytorch/pull/33019) into two parts. As per our discussion with @eellison,  the first part is to make sure our aten::slice operator take optional parameters for begin/step/end. This will help with refactoring ir_emitter.cpp for genering handling for list and slice striding. Once this PR merged, we will submit a second PR with compiler change. 

Test Plan: None for this PR, but new tests will be added for the second part.

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Striding for lists Part 1"


Summary: Attempt to break this PR (https://github.com/pytorch/pytorch/pull/33019) into two parts. As per our discussion with @eellison,  the first part is to make sure our aten::slice operator take optional parameters for begin/step/end. This will help with refactoring ir_emitter.cpp for genering handling for list and slice striding. Once this PR merged, we will submit a second PR with compiler change. 

Test Plan: None for this PR, but new tests will be added for the second part.

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

eellison commented:
cc @ailzhang XLA failure

eellison commented:
Some context: we are fixing the lists slicing logic with a negative index. Part of this change involves changing the ir emission for aten::slice to emit None for non-present step value instead of INT_MAX. Rather than have special case logic for all of the builtins type that may or may not support negative striding (lists, tensors, strs) we are updating their implementations so that they can all take in optionals. Actual negative slicing of Tensors is still NYI, but making the schema changes here makes future implementation easier and makes the logic in ir emission cleaner.

tugsbayasgalan committed:
Update on "Striding for lists Part 1"


Summary: Attempt to break this PR (https://github.com/pytorch/pytorch/pull/33019) into two parts. As per our discussion with @eellison,  the first part is to make sure our aten::slice operator take optional parameters for begin/step/end. This will help with refactoring ir_emitter.cpp for genering handling for list and slice striding. Once this PR merged, we will submit a second PR with compiler change. 

Test Plan: None for this PR, but new tests will be added for the second part.

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Striding for lists Part 1"


Summary: Attempt to break this PR (https://github.com/pytorch/pytorch/pull/33019) into two parts. As per our discussion with @eellison,  the first part is to make sure our aten::slice operator take optional parameters for begin/step/end. This will help with refactoring ir_emitter.cpp for genering handling for list and slice striding. Once this PR merged, we will submit a second PR with compiler change. 

Test Plan: None for this PR, but new tests will be added for the second part.

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

eellison commented on aten/src/ATen/BatchingRegistrations.cpp:
```diff
@@ -340,7 +340,12 @@ Tensor select_backward_batching_rule(const Tensor& grad, IntArrayRef input_sizes
 }
 
-Tensor slice_batching_rule(const Tensor& self, int64_t dim, int64_t start, int64_t end, int64_t step) {
+Tensor slice_batching_rule(
```
cc: @zou3519 , i think these changes are harmless but double checking

zou3519 replied:
Yup, this lgtm

eellison commented on aten/src/ATen/native/TensorShape.cpp:
```diff
@@ -1070,36 +1070,52 @@ Tensor index_select_sparse(const Tensor& self, int64_t dim, const Tensor& index)
+    int64_t dim,
+    c10::optional<int64_t> start,
+    c10::optional<int64_t> end,
+    c10::optional<int64_t> step) {
```
In python, `[::]` corresponds to None, None, 1 right ? I don't think we actually need to make step an optional

eellison commented on aten/src/ATen/native/TensorShape.cpp:
```diff
@@ -1070,36 +1070,52 @@ Tensor index_select_sparse(const Tensor& self, int64_t dim, const Tensor& index)
+  int64_t step_val = step.has_value() ? step.value() : 1;
+
   // TODO: support negative strides
-  TORCH_CHECK(step > 0, "slice step must be positive");
```
Why were these lines changed ? I think we should leave negative indexing NYI for tensors, it might be tricky, and i think is out of scope for this PR. Just error'ing shoudl be fine.

tugsbayasgalan replied:
I think it was just clang formatting

eellison replied:
oh, nvm then my bad. Changes heree look good 👍

eellison reviewed (commented):
Looks great! Nice job wading through the tracer code.

I have two comments:
- I don't think we actually need to make `step` optional, I think python just defaults it to 1
- I'm not sure about the `TenssorShape.cpp` changes, I think it would be better to error at negative step within TensorShape instead of trying to implement it for this PR.

eellison commented on aten/src/ATen/native/TensorShape.cpp:
```diff
@@ -1070,36 +1070,52 @@ Tensor index_select_sparse(const Tensor& self, int64_t dim, const Tensor& index)
   auto sizes = self.sizes().vec();
   auto strides = self.strides().vec();
+
+  // handle optinal parameters
```
nit: /s/optinal/optional

eellison replied:
bump lol

tugsbayasgalan committed:
Update on "Striding for lists Part 1"


Summary: Attempt to break this PR (https://github.com/pytorch/pytorch/pull/33019) into two parts. As per our discussion with @eellison,  the first part is to make sure our aten::slice operator take optional parameters for begin/step/end. This will help with refactoring ir_emitter.cpp for genering handling for list and slice striding. Once this PR merged, we will submit a second PR with compiler change. 

Test Plan: None for this PR, but new tests will be added for the second part.

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Striding for lists Part 1"


Summary: Attempt to break this PR (https://github.com/pytorch/pytorch/pull/33019) into two parts. As per our discussion with @eellison,  the first part is to make sure our aten::slice operator take optional parameters for begin/step/end. This will help with refactoring ir_emitter.cpp for genering handling for list and slice striding. Once this PR merged, we will submit a second PR with compiler change. 

Test Plan: None for this PR, but new tests will be added for the second part.

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Striding for lists Part 1"


Summary: Attempt to break this PR (https://github.com/pytorch/pytorch/pull/33019) into two parts. As per our discussion with @eellison,  the first part is to make sure our aten::slice operator take optional parameters for begin/step/end. This will help with refactoring ir_emitter.cpp for genering handling for list and slice striding. Once this PR merged, we will submit a second PR with compiler change. 

Test Plan: None for this PR, but new tests will be added for the second part.

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Striding for lists Part 1"


Summary: Attempt to break this PR (https://github.com/pytorch/pytorch/pull/33019) into two parts. As per our discussion with @eellison,  the first part is to make sure our aten::slice operator take optional parameters for begin/step/end. This will help with refactoring ir_emitter.cpp for genering handling for list and slice striding. Once this PR merged, we will submit a second PR with compiler change. 

Test Plan: None for this PR, but new tests will be added for the second part.

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

eellison reviewed (approved):
LGTM, Ship it! Great job! We might want to sanity check the same forward-compatibility checks we did previously with the newest changes, just to be safe. But up to you.

eellison commented on torch/csrc/jit/runtime/register_prim_ops.cpp:
```diff
@@ -29,23 +30,25 @@ namespace {
 
 std::string stringSlice(
```
Actually, sorry, one more comment: just as we updated the list implementation to work with negative step we should do the same with strings right ?

tugsbayasgalan committed:
Update on "Striding for lists Part 1"


Summary: Attempt to break this PR (https://github.com/pytorch/pytorch/pull/33019) into two parts. As per our discussion with @eellison,  the first part is to make sure our aten::slice operator take optional parameters for begin/step/end. This will help with refactoring ir_emitter.cpp for genering handling for list and slice striding. Once this PR merged, we will submit a second PR with compiler change. 

Test Plan: None for this PR, but new tests will be added for the second part.

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Striding for lists Part 1"


Summary: Attempt to break this PR (https://github.com/pytorch/pytorch/pull/33019) into two parts. As per our discussion with @eellison,  the first part is to make sure our aten::slice operator take optional parameters for begin/step/end. This will help with refactoring ir_emitter.cpp for genering handling for list and slice striding. Once this PR merged, we will submit a second PR with compiler change. 

Test Plan: None for this PR, but new tests will be added for the second part.

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

eellison reviewed (approved):
LGTM, ship it! We should wait a couple weeks or so before landing PT2 so we dont run into FC issues.

tugsbayasgalan committed:
Update on "Striding for lists Part 1"


Summary: Attempt to break this PR (https://github.com/pytorch/pytorch/pull/33019) into two parts. As per our discussion with @eellison,  the first part is to make sure our aten::slice operator take optional parameters for begin/step/end. This will help with refactoring ir_emitter.cpp for genering handling for list and slice striding. Once this PR merged, we will submit a second PR with compiler change. 

Test Plan: None for this PR, but new tests will be added for the second part.

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Striding for lists Part 1"


Summary: Attempt to break this PR (https://github.com/pytorch/pytorch/pull/33019) into two parts. As per our discussion with @eellison,  the first part is to make sure our aten::slice operator take optional parameters for begin/step/end. This will help with refactoring ir_emitter.cpp for genering handling for list and slice striding. Once this PR merged, we will submit a second PR with compiler change. 

Test Plan: None for this PR, but new tests will be added for the second part.

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Striding for lists Part 1"


Summary: Attempt to break this PR (https://github.com/pytorch/pytorch/pull/33019) into two parts. As per our discussion with @eellison,  the first part is to make sure our aten::slice operator take optional parameters for begin/step/end. This will help with refactoring ir_emitter.cpp for genering handling for list and slice striding. Once this PR merged, we will submit a second PR with compiler change. 

Test Plan: None for this PR, but new tests will be added for the second part.

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D25929902](https://our.internmc.facebook.com/intern/diff/D25929902)

[ghstack-poisoned]

ezyang commented on torch/csrc/jit/runtime/register_prim_ops.cpp:
```diff
@@ -29,23 +31,22 @@ namespace {
-  TORCH_CHECK(step == 1, "Slicing a string only supports step=1");
-
-  const int64_t size = string.size();
+  int64_t start_val = start.has_value() ? start.value() : INT64_MAX;
```
Why does this default to `INT64_MAX`? Why not just set it to 0? cc @eellison this is impeding SymInt-ification

eellison replied:
Ugh, I forget, but I think that was mirroring how the python slice handling code went where `None` would be encoded as `INT64_MAX`
